examples/tv-casting-app/darwin/MatterTvCastingBridge/MatterTvCastingBridge/CastingServerBridge.mm

```diff
@@ -118,10 +109,157 @@ - (instancetype)init
+    // is returning to the _chipWorkQueue thread; this can cause callbacks to be received in an unexpected
+    // order, where the result might arrive before the client knows that the call was started.
+    //
+    // To avoid this confusion, we perform a "double async" dispatch to ensure that events are delivered in a
+    // sensible order. First, we schedule the result callback to run on the _chipWorkQueue (so it *must* run
+    // after the end of the CastingServerBridge's wrapper logic that already started on the _chipWorkQueue)
+    // and then immediately, asynchronously dispatch the callback onto the provided client queue.
+    dispatch_async(_chipWorkQueue, ^{
+        if (nil != description) {
+            ChipLogProgress(AppServer, "[ASYNC (CHIP)] CastingServerBridge dispatching %s", [description UTF8String]);
+        }
+        dispatch_async(queue, ^{
+            if (nil != description) {
+                ChipLogProgress(AppServer, "[ASYNC (CLIENT)] CastingServerBridge BEGIN %s", [description UTF8String]);
+            }
+            block();
+            if (nil != description) {
+                ChipLogProgress(AppServer, "[ASYNC (CLIENT)] CastingServerBridge END %s", [description UTF8String]);
+            }
+        });
+    });
+}
+
+/**
+ A utility function that invokes an Objective C Block on the Matter Dispatch Queue in a thread-safe manner.
+
+ The result of the Objective C Block will be relayed to the `blockCompleteCallback`.
+
+ @param description  Optional. A description of the Objective C Block to be invoked for diagnostic purposes.
+ @param blockCompleteCallback The callback to be invoked with the result of the Objective C Block.
+ @param callbackQueue The Dispatch Queue on which the callback shall be invoked.
+ @param block The Objective C Block to dispatch.
+ */
+- (void) withCastingServerInvokeBlock:(const NSString * _Nullable)description
+               withCompletionCallback:(void (^_Nonnull)(bool))blockCompleteCallback
+                              onQueue:(dispatch_queue_t)callbackQueue
+                                block:(CHIP_ERROR (^_Nonnull)(CastingServer *)) block
+{
+    [self dispatchOnMatterSDKQueue:description block:^{
+        CastingServer * castingServer = CastingServer::GetInstance();
+        const CHIP_ERROR result = block(castingServer);
+        dispatch_async(callbackQueue, ^{
+            if (nil != description) {
+                ChipLogProgress(AppServer, "[ASYNC (CLIENT)] CastingServerBridge BEGIN %s Started Callback", [description UTF8String]);
+            }
+            blockCompleteCallback([[MatterError alloc] initWithCode:result.AsInteger()
+                                                      message:[NSString stringWithUTF8String:result.AsString()]]);
+            if (nil != description) {
+                ChipLogProgress(AppServer, "[ASYNC (CLIENT)] CastingServerBridge END %s Started Callback", [description UTF8String]);
```

Comment:
Should this be Completed Callback?

```suggestion
                ChipLogProgress(AppServer, "[ASYNC (CLIENT)] CastingServerBridge END %s Completed Callback", [description UTF8String]);
```

Reply:
It's a little confusing, but the `blockCompleteCallback` is invoked to inform the client that the Matter SDK has been asynchronously invoked (started) and to expect the result callback or subscription callbacks to be called next. The log statement helps with tracing the execution of the client's SDK-operation-started-callback.

Reply:
Yeah its weird to have both START / END for a "Started" callback, i wasn't sure how to interpret it.

Can we do something like "About to start Callback", "Callback started" or something to make it more clear?

Reply:
I changed it to `CastingServerBridge invoking %s Started Callback` and `CastingServerBridge invoked %s Started Callback`, I think that reads OK.